Add tests to improve coverage

vladimirfomene opened the pull request:
### Description

This PR add more test to the database module and also fixes certain bugs discovered by the written test. I also amended the name used for the database parameter in the test functions.

### Notes to the reviewers

This contributes to fixing bitcoindevkit/bdk_wallet#179 

### Changelog notice

<!-- Notice the release manager should include in the release tag message changelog -->
<!-- See https://keepachangelog.com/en/1.0.0/ for examples -->

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

danielabrozzoni commented:
I'd like to see b91afdb49bcec89007b254883fa3c21420e2f382 split into multiple commits: at least one different commit for each bug discovered and fixed, with an explanation of the bug found. You modified quite a bit of code in `src/database/keyvalue.rs` and `src/database/memory.rs`, but without a clear commit message it's nearly to impossible to understand your intentions.

vladimirfomene committed:
Add more test to the database module

This PR aims to add more test to database
code so that we can catch bugs as soon
as they occur. Contributing to fixing
issue #699.

vladimirfomene commented:
@danielabrozzoni, I have implemented the changes requested and explained reason for the change in the commit message.

danielabrozzoni commented on src/database/mod.rs:
```diff
@@ -441,5 +458,180 @@ pub mod test {
+        let (chain, child) = db.del_path_from_script_pubkey(&script).unwrap().unwrap();
+
+        assert_eq!(chain, keychain);
+        assert_eq!(path, child);
```
I think you should also iter script pub keys to check if they were actually deleted from the db
(and do the same for the others `del_` tests)

danielabrozzoni commented on src/database/mod.rs:
```diff
@@ -441,5 +458,180 @@ pub mod test {
+        assert_eq!(res, 1);
+    }
+
+    pub fn test_check_descriptor_checksum<D: Database>(mut db: D) {
```
I think you should add a couple of comments here to make the test more understandable... If I understand correctly, the first check_descriptor_checksum sets the checksum to "1cead456", and the second one compares the stored checksum with "1cead454", and returns an error. Does that make sense?

vladimirfomene replied:
Yes, your thinking about the logic is right. I will add comments here.

danielabrozzoni reviewed (commented):
First of all, some styling nits:
1. You shouldn't put "Fix failing test: test_del_tx" as the commit message for 27c6735d122ea7ef61f19e1eeac23643ccc096e8 - your commit is not fixing a test. Your commit is fixing a bug you've found. This is very different, as "Fix failing test" makes the reader think that in the commit you are modifying the actual test behavior.
2. This is a quite big nit, but I usually first put commits that fix bugs and then commits that add tests, not the other way around (in this PR d8f502660f30327f5ffe1ef6a961641bdfad539e is before 27c6735d122ea7ef61f19e1eeac23643ccc096e8). The reason for that is: it helps me write a clearer commit message, as that I'm not tempted to cite any test in the bug-fixing commit, but also it makes the git history bisectable, which is always a nice to have.

The added tests look good and they actually improve the code coverage :) See https://coveralls.io/builds/52344495

I am a bit confused about 27c6735d122ea7ef61f19e1eeac23643ccc096e8: to me, it doesn't seem that you need to modify `select_transaction_details_by_txid`, you can just remove the transaction from  the returned object if `include_raw` is false:
```
diff --git a/src/database/sqlite.rs b/src/database/sqlite.rs
index 562c934..4fb6337 100644
--- a/src/database/sqlite.rs
+++ b/src/database/sqlite.rs
@@ -744,11 +744,16 @@ impl BatchOperations for SqliteDatabase {
         include_raw: bool,
     ) -> Result<Option<TransactionDetails>, Error> {
         match self.select_transaction_details_by_txid(txid)? {
-            Some(transaction_details) => {
+            Some(mut transaction_details) => {
                 self.delete_transaction_details_by_txid(txid)?;
 
                 if include_raw {
                     self.delete_transaction_by_txid(txid)?;
+                } else {
+                    // Since we want to return what we actually deleted from
+                    // the database, we make sure to put a "None" in the
+                    // transaction field here.
+                    transaction_details.transaction = None;
                 }
                 Ok(Some(transaction_details))
             }
```

Is there something I'm not seeing?

vladimirfomene commented:
@danielabrozzoni, initially I made the transaction field `None` on the returned object and yes it fixed the bug. The other issue here is that if I was calling `get_tx` with `include_raw = false`, behind the scene `select_transaction_details_by_txid` will get the transaction even though it is not needed as it does a join query on `transactions` and `transactiondetails`.

afilini commented on src/database/sqlite.rs:
```diff
@@ -467,7 +467,7 @@ impl SqliteDatabase {
         &self,
         txid: &[u8],
     ) -> Result<Option<TransactionDetails>, Error> {
-        let mut statement = self.connection.prepare_cached("SELECT transaction_details.timestamp, transaction_details.received, transaction_details.sent, transaction_details.fee, transaction_details.height, transactions.raw_tx FROM transaction_details, transactions WHERE transaction_details.txid=transactions.txid AND transaction_details.txid=:txid")?;
```
I would keep this query the same as it was before, because while it doesn't make too much difference for sqlite I think it would be a lot slower to make two queries over network rather than one. So if in the future we add support for postgres or mysql it will probably be based on the sqlite code, and for this reason I would try to be forward-thinking here.

danielabrozzoni replied:
If it makes the code cleaner, we might decide to pass a `include_raw` parameter to `select_transaction_details_by_txid` and, even though we will still fetch the transaction from the db using this query, decide not to return it if it's not needed

vladimirfomene replied:
I will go with just setting the transaction field of the return object to None.

vladimirfomene committed:
Set tx field to none if `include_raw` is false

`del_tx` pulls the TransactionDetails object using
`select_transaction_details_by_txid` method which gets the transaction
details' data with a non-None transaction field even if the
`include_raw` argument is `false`. So it becomes necessary to Set
the transaction field in transactiondetails to None in `del_tx`, when
we make a call to it with `include_raw=false`.

danielabrozzoni commented on src/database/mod.rs:
```diff
@@ -619,13 +631,21 @@ pub mod test {
         let keychain = KeychainKind::External;
         let _res = db.check_descriptor_checksum(keychain, checksum);
 
+        // check if `check_descriptor_checksum` throws
```
Those comments are great, but included in the wrong commit - should be in fd48469b617b2a0335d8a982d913434e8b6df77e, not 2ea1f574928b85ee00829efa9cfc4b23e8a712df

danielabrozzoni commented on src/database/mod.rs:
```diff
@@ -494,7 +491,10 @@ pub mod test {
 
         assert_eq!(chain, keychain);
-        assert_eq!(path, child);
+        assert_eq!(child, path);
+
+        let res = db.get_path_from_script_pubkey(&script).unwrap();
+        assert!(res.is_none());
```
Same for this test additions (and more below): should be in https://github.com/bitcoindevkit/bdk/commit/fd48469b617b2a0335d8a982d913434e8b6df77e, not https://github.com/bitcoindevkit/bdk/commit/2ea1f574928b85ee00829efa9cfc4b23e8a712df

vladimirfomene committed:
Change parameter name of database in test funcs

Change parameter name in database test functions
from `tree` to `db`.

rajarshimaitra reviewed (approved):
tACK e65edbf53cb14e273ff99ed0576843dd3367385d

Thanks for the new tests..

danielabrozzoni reviewed (approved):
Code review ACK e65edbf53cb14e273ff99ed0576843dd3367385d
